### PR TITLE
GitHub pull request tools

### PR DESCRIPTION
**Pull request by TransformerOptimus**
### Description

Github fetch pull requests and review pull request tools

### Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs update

### Checklist
- [X] My pull request is atomic and focuses on a single change.
- [ ] I have read the contributing guide and my code conforms to the guidelines.
- [ ] I have documented my changes clearly and comprehensively.
- [ ] I have added the required tests.

### CODE REVIEW

File: superagi/llms/llm_model_factory.py [hunk 1]
```diff
@@ -33,5 +33,21 @@ def get_model(organisation_id, api_key, model="gpt-3.5-turbo", **kwargs):
     elif provider_name == 'Hugging Face':
         print("Provider is Hugging Face")
         return HuggingFace(model=model_instance.model_name, end_point=model_instance.end_point, api_key=api_key, **kwargs)
+    else:
+        print('Unknown provider.')
+
+def build_model_with_api_key(provider_name, api_key):
+    if provider_name.lower() == 'openai':
+        print("Provider is OpenAI")
```
Comment (luciferlinx101) [inline superagi/llms/llm_model_factory.py:41]: print statements can be removed

File: superagi/llms/llm_model_factory.py [hunk 2]
```diff
@@ -33,5 +33,21 @@ def get_model(organisation_id, api_key, model="gpt-3.5-turbo", **kwargs):
     elif provider_name == 'Hugging Face':
         print("Provider is Hugging Face")
         return HuggingFace(model=model_instance.model_name, end_point=model_instance.end_point, api_key=api_key, **kwargs)
+    else:
+        print('Unknown provider.')
```
Comment (luciferlinx101) [inline superagi/llms/llm_model_factory.py:37]: print statements can be removed and logger can be used

File: superagi/llms/llm_model_factory.py [hunk 3]
```diff
@@ -33,5 +33,21 @@ def get_model(organisation_id, api_key, model="gpt-3.5-turbo", **kwargs):
...
+
+def build_model_with_api_key(provider_name, api_key):
+    if provider_name.lower() == 'openai':
+        print("Provider is OpenAI")
+        return OpenAi(api_key=api_key)
+    elif provider_name.lower() == 'replicate':
+        print("Provider is Replicate")
```
Comment (luciferlinx101) [inline superagi/llms/llm_model_factory.py:44]: print statements can be removed and logger can be used

File: superagi/llms/llm_model_factory.py [hunk 4]
```diff
@@ -33,5 +33,21 @@ def get_model(organisation_id, api_key, model="gpt-3.5-turbo", **kwargs):
...
+        return Replicate(api_key=api_key)
+    elif provider_name.lower() == 'google palm':
+        print("Provider is Google Palm")
```
Comment (luciferlinx101) [inline superagi/llms/llm_model_factory.py:47]: print statements can be removed and logger can be used

File: superagi/llms/llm_model_factory.py [hunk 5]
```diff
@@ -33,5 +33,21 @@ def get_model(organisation_id, api_key, model="gpt-3.5-turbo", **kwargs):
...
+        return GooglePalm(api_key=api_key)
+    elif provider_name.lower() == 'hugging face':
+        print("Provider is Hugging Face")
```
Comment (luciferlinx101) [inline superagi/llms/llm_model_factory.py:50]: print statements can be removed and logger can be used

File: superagi/helper/github_helper.py [hunk 6]
```diff
@@ -335,3 +337,112 @@ def _get_file_contents(self, file_name, agent_id, agent_execution_id, session):
...
+        # Search query
+        query = f'repo:{repository_owner}/{repository_name} type:pr created:>{time_x_seconds_ago_str}'
+
+        url = f'https://api.github.com/search/issues?q={query}'
+        headers = {
+            "Authorization": f"token {self.github_access_token}",
+            "Content-Type": "application/json",
+        }
+
+        response = requests.get(url, headers=headers)
```
Comment (luciferlinx101) [inline superagi/helper/github_helper.py:439]: here we can avoid using url and headers and make method small by passing them as param and avoid url,headers as temp var?

File: superagi/helper/github_helper.py [hunk 7]
```diff
@@ -335,3 +337,112 @@ def _get_file_contents(self, file_name, agent_id, agent_execution_id, session):
...
+            "Accept": "application/vnd.github.v3+json"
+        }
+        data = {
+            "commit_id": commit_id,
+            "path": file_path,
+            "position": position,
+            "body": comment_body
+        }
+        response = requests.post(comments_url, headers=headers, json=data)
+        if response.status_code == 201:
```
Comment (luciferlinx101) [inline superagi/helper/github_helper.py:405]: here we can avoid using url and headers and make method small by passing them as param and avoid url,headers,data as temp var?

File: superagi/llms/llm_model_factory.py [hunk 8]
```diff
@@ -33,5 +33,21 @@ def get_model(organisation_id, api_key, model="gpt-3.5-turbo", **kwargs):
...
+        return HuggingFace(api_key=api_key)
     else:
         print('Unknown provider.')
```
Comment (luciferlinx101) [inline superagi/llms/llm_model_factory.py:53]: Print statement can be changed to logger

File: superagi/helper/github_helper.py [hunk 9]
```diff
@@ -9,7 +9,8 @@
 from superagi.types.storage_types import StorageType
```
Comment (luciferlinx101) [inline superagi/helper/github_helper.py:9]: Can we now on put a helper in the tool folder making a new folder like inside GitHub a folder named helper and inside that all helper logic as we might need to extract out all tools in the marketplace tool repo?

File: superagi/llms/llm_model_factory.py [hunk 10]
```diff
@@ -33,5 +33,21 @@ def get_model(organisation_id, api_key, model="gpt-3.5-turbo", **kwargs):
     elif provider_name == 'Hugging Face':
         print("Provider is Hugging Face")
```
Comment (luciferlinx101) [inline superagi/llms/llm_model_factory.py:34]: print can be removed to logger

File: tests/unit_tests/tools/github/test_review_pull_request.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    assert result == expected
+
+
+@pytest.mark.parametrize("diff_content, file_path, line_number, expected", [
+    ("file_path_1\n@@ -1,3 +1,4 @@\n+ line1\n+ line2\n+ line3", "file_path_1", 3, 4),
+    ("file_path_2\n@@ -1,3 +1,3 @@\n+ line1\n- line2", "file_path_2", 1, 2),
+    ("file_path_3\n@@ -1,3 +1,4 @@\n+ line1\n+ line2\n- line3", "file_path_3", 2, 3)
+])
+def test_get_exact_line_number(diff_content, file_path, line_number, expected):
+    print(file_path)
```
Comment (luciferlinx101) [inline tests/unit_tests/tools/github/test_review_pull_request.py:42]: remove print statements

File: superagi/tools/github/review_pull_request.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+    pull_request_number: int = Field(
+        ...,
+        description="Pull request number",
+    )
+
+
+class GithubReviewPullRequest(BaseTool):
+    """
+    Reviews the githun pull request and adds comments inline
+
```
Comment (luciferlinx101) [inline superagi/tools/github/review_pull_request.py:33]: Typo githun -> github

File: superagi/tools/github/review_pull_request.py [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+            pull_request_number: pull request number
+
+        Returns:
+            Pull request success message if pull request is created successfully else error message.
+        """
+        try:
+            github_access_token = self.get_tool_config("GITHUB_ACCESS_TOKEN")
+            github_username = self.get_tool_config("GITHUB_USERNAME")
+            github_helper = GithubHelper(github_access_token, github_username)
+
```
Comment (luciferlinx101) [inline superagi/tools/github/review_pull_request.py:62]: we can avoid the use of github_helper as a temporary variable

File: superagi/llms/llm_model_factory.py [hunk 14]
```diff
@@ -33,5 +33,21 @@ def get_model(organisation_id, api_key, model="gpt-3.5-turbo", **kwargs):
...
         return HuggingFace(model=model_instance.model_name, end_point=model_instance.end_point, api_key=api_key, **kwargs)
+    else:
+        print('Unknown provider.')
+
+def build_model_with_api_key(provider_name, api_key):
```
Comment (luciferlinx101) [inline superagi/llms/llm_model_factory.py:39]: Unit Test for build_model_with_api_key can be added

File: superagi/helper/github_helper.py [hunk 15]
```diff
@@ -335,3 +337,112 @@ def _get_file_contents(self, file_name, agent_id, agent_execution_id, session):
...
+            logger.info('Successfully fetched pull request content.')
+            return response.text
+        elif response.status_code == 404:
+            logger.warning('Pull request not found.')
+        else:
+            logger.warning('Failed to fetch pull request content: ', response.text)
+
+        return None
+
+    def get_latest_commit_id_of_pull_request(self, repository_owner, repository_name, pull_request_number):
```
Comment (luciferlinx101) [inline superagi/helper/github_helper.py:374]: Needs code docs and comment

File: superagi/helper/github_helper.py [hunk 16]
```diff
@@ -335,3 +337,112 @@ def _get_file_contents(self, file_name, agent_id, agent_execution_id, session):
...
+        if response.status_code == 200:
+            commits = response.json()
+            latest_commit = commits[-1]  # Assuming the last commit is the latest
+            return latest_commit.get('sha')
+        else:
+            logger.warning(f'Failed to fetch commits for pull request: {response.json()["message"]}')
+            return None
+
+
+    def add_line_comment_to_pull_request(self, repository_owner, repository_name, pull_request_number,
```
Comment (luciferlinx101) [inline superagi/helper/github_helper.py:390]: Needs code docs and comments

File: superagi/tools/github/prompts/code_review.txt [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+Following is the github pull request diff content:
+```
+{{DIFF_CONTENT}}
+```
+
+Instructions:
+1. Do not comment on existing lines and deleted lines.
+2. Ignore the lines start with '-'.
+3. Only consider lines starting with '+' for review.
+3. Do not comment on frontend and graphql code.
```
Comment (luciferlinx101) [inline superagi/tools/github/prompts/code_review.txt:18]: Typo : It should be 3. - > 4.

File: superagi/tools/github/review_pull_request.py (hunk 13 above)
Comment (TransformerOptimus) [reply superagi/tools/github/review_pull_request.py:62]: We need to pass it to next method.

**Pull request merged**